AMD ROCm 3.10 Release: Add new package recipe for rocm-openmp-extras.

estewart08 opened the pull request.

estewart08 committed:
AMD ROCm 3.9 Release: Add new package recipe for openmp-extras.

estewart08 commented:
Looks like I cannot add a reviewer. Would like @adamjstewart to review. AMD label will be added.

adamjstewart commented on var/spack/repos/builtin/packages/openmp-extras/package.py:
```diff
@@ -0,0 +1,319 @@
+
+    depends_on('cmake@3.5.2:3.13.4', type='build')
+    depends_on('mesa18~llvm@18.3:', type=('build', 'link'))
+    depends_on('py-setuptools@44.1.0', type='build')
+    depends_on('python@3.8.6', type='build', when='@3.9.0')
```
You should use version ranges for these. For example, if Python 3.5+ is known to work, then use `python@3.5:` with a trailing colon. Also, the `when` isn't needed since you only have one version.

estewart08 replied:
done

adamjstewart commented on var/spack/repos/builtin/packages/openmp-extras/package.py:
```diff
@@ -0,0 +1,319 @@
+        ]
+
+        # Override standard CMAKE_BUILD_TYPE
+        std_cmake_args.remove("-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo")
```
All cmake packages have a `build_type=RelWithDebInfo` variant. If someone changes this variant on the command line, it will no longer be RelWithDebInfo, so this isn't very robust. I guess you'll need to loop through and remove any args containing the string `CMAKE_BUILD_TYPE`.

estewart08 replied:
done

adamjstewart reviewed (commented):
@haampie might want to review this

There seems to be a lot of overlap between this package and LLVM, so pinging @naromero77 @trws

naromero77 commented:
Is this package in support of the AMD compiler?

alalazo commented on var/spack/repos/builtin/packages/openmp-extras/package.py:
```diff
@@ -0,0 +1,319 @@
+}
+
+
+class OpenmpExtras(Package):
+    """OpenMP support for ROCm LLVM."""
```
The name of the package feels a bit too generic...

estewart08 replied:
I can understand that. We could possibly change the package name to AmdgpuOpenmpExtras or RocmOpenmpExtras.

estewart08 replied:
Changed to rocm-openmp-extras.

estewart08 commented:
Thanks for the feedback. I had a long response/explanation written, but need to look into one issue further so I can provide more clarity for you on the llvm overlap here.

estewart08 committed:
Updated openmp-extras recipe.

Changed name to rocm-openmp-extras.
Reduced repitition with component hashes.
Adjusted the removal of CMAKE_BUILD_TYPE.
Reduced repitition in cmake args.
Added 3.10.0 to recipe.

estewart08 committed:
Update python/py-setuptools dependency versions.

estewart08 commented:
Updated the recipe with suggested changes, see commits for details. I was able to remove the re-build of llvm-amdgpu. 
Added 3.10.0 to the recipe.

adamjstewart commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -0,0 +1,337 @@
+    depends_on('cmake@3.5.2:3.13.4', type='build')
+    depends_on('mesa18~llvm@18.3:', type=('build', 'link'))
+    depends_on('py-setuptools@44.1.0:', type='build')
+    depends_on('python@3.6.9:3.8.6', type='build')
```
It won't work with Python 3.6.8 or 3.8.7?

adamjstewart commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -0,0 +1,337 @@
+    version('3.10.0', sha256=versions_dict['3.10.0']['aomp'])
+    version('3.9.0', sha256=versions_dict['3.9.0']['aomp'])
+
+    depends_on('cmake@3.5.2:3.13.4', type='build')
```
It won't build with newer CMake?

estewart08 replied:
It very well could. Is it expected of a recipe to test every possible version combination of its dependencies? If so, would this range suffice for now?

adamjstewart replied:
It's not expected. If you don't know, I wouldn't add an upper limit. The lower limit generally comes from the `CMakeLists.txt`.

estewart08 replied:
In my opinion, removing the upper limit has the potential of breaking the recipe for someone.  I would rather have a recipe that takes a bit longer to build over a non-functional one, but I can definitely see the speed benefit if the newer version does work.

adamjstewart commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -0,0 +1,337 @@
+            libomptarget.format(src) + '/plugins/hsa/CMakeLists.txt')
+
+        filter_file(
+            r'-L\${LLVM_LIBDIR}\${OPENMP_LIBDIR_SUFFIX}',
```
If you use `string=True`, you won't need to escape special chars because it won't treat it as a regex.

estewart08 replied:
Noted. This can be addressed in the next version update.

estewart08 replied:
Decided to go ahead and push this change.

adamjstewart commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -0,0 +1,337 @@
+            '-DDEVICE_LIBS_DIR={0}/amdgcn/bitcode'.format(devlibs_prefix),
+            '-DAOMP_STANDALONE_BUILD=0',
+            '-DDEVICELIBS_ROOT={0}'.format(devlibs_src),
+            '-DOPENMP_TEST_C_COMPILER=$AOMP/bin/clang',
+            '-DOPENMP_TEST_CXX_COMPILER=$AOMP/bin/clang++',
```
These won't expand to the value of `AOMP` like they would in bash, just so you know.

estewart08 committed:
Expand dependency versions, fix the openmp cmake args.

adamjstewart commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -0,0 +1,337 @@
+        devlibs_prefix = self.spec['rocm-device-libs'].prefix
+        openmp_extras_prefix = self.spec['rocm-openmp-extras'].prefix
+        llvm_prefix = self.spec['llvm-amdgpu'].prefix
+        env.set('AOMP', '{0}'.format(format(llvm_prefix)))
```
I don't think all of these double `format`s do anything, you only need a single `format`

estewart08 replied:
Indeed, this was overlooked.

adamjstewart commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -0,0 +1,337 @@
+        env.set('HIP_DEVICE_LIB_PATH',
+                '{0}/amdgcn/bitcode'.format(format(devlibs_prefix)))
+        env.set('AOMP_GPU',
+                '`{0}/rocm-bin/mygpu`'.format(format(openmp_extras_prefix)))
```
I assume you don't want backticks in this variable

estewart08 replied:
These are intended and will set this variable to what gpu is determined and returned from mygpu. Our test suites leverage this variable.

adamjstewart replied:
I'm still not sure I understand. If you run these on the command line, bash will execute the `mygpu` command. But since we aren't using bash, the entire string, including the backticks, will be passed to CMake. Is that what you want?

estewart08 replied:
The AOMP_GPU variable is intended to be used in bash when users run our examples.  I am choosing to automate this in the script to configure a user's non-build environment for convenience.  This is not intended for cmake usage.  Is using ```setup_run_environment``` not the proper way to achieve this?

adamjstewart replied:
Hmm, `setup_run_environment` is the correct way to set environment variables, but to my best knowledge, this will end up being equivalent to:
```bash
export AOMP_GPU='`/path/to/rocm-bin/mygpu`'
```
Spack won't interpret the backticks, and neither will bash. If you're trying to run that command and store the output in `AOMP_GPU`, you'll want to use `Executable` instead.

estewart08 replied:
I am not familiar with ```Executable``` in Spack.

I might need to clarify a little further. The AOMP_GPU variable actually is evaluated in the Makefile of our examples as 
```
-march=$(AOMP_GPU)
``` 
which is passed on to clang. On my machine, I can verify that this returns -march=gfx906. 

If these backticks are still a problem, I could resolve with some python code instead:
```
my_gpu = subprocess.check_output(openmp_extras_prefix + '/rocm-bin/mygpu').decode(sys.stdout.encoding)
env.set('AOMP_GPU', my_gpu)
```

adamjstewart replied:
Ah, I see. As long as the Makefile executes the backticks, then it's fine. Thanks for clarifying!

estewart08 committed:
Remove double formats.

estewart08 committed:
Substitute escape characters with string=True for filter_file(s).

adamjstewart commented on var/spack/repos/builtin/packages/rocm-openmp-extras/package.py:
```diff
@@ -0,0 +1,342 @@
+            '-DDEVICELIBS_ROOT={0}'.format(devlibs_src),
+            '-DNEW_BC_PATH=1',
+            '-DAOMP={0}'.format(llvm_prefix),
+            '-DCMAKE_VERBOSE_MAKEFILE=1'
```
This is already part of `std_cmake_args`, you don't need to add this

estewart08 replied:
done

estewart08 committed:
Remove CMAKE_VERBOSE_MAKEFILE, already present in standard args.

arjun-raj-kuppala commented:
@adamjstewart @haampie Can you help getting this approved/merged?

anything pending on this ?

adamjstewart commented:
See my comment on `AOMP_GPU` above, I'm still not sure about the use of backticks.

estewart08 commented:
This all set to merge now?
